gogen/codegen.go

```diff
@@ -0,0 +1,429 @@
+	EnumTypeMap string
+}
+
+// NewGoCodeGenerator returns a new instance of the GoCodeGenerator
+// struct to the calling function.
+func NewGoCodeGenerator(c *ygen.GeneratorConfig, goopts *GoOpts) *GoCodeGenerator {
+	cg := &GoCodeGenerator{}
+
+	if c != nil {
+		cg.Config = *c
+	}
+	if goopts != nil {
+		cg.GoOptions = *goopts
+	}
+
+	return cg
+}
+
+// checkForBinaryKeys returns a non-empty list of errors if the input directory
+// has one or more binary types (including union types containing binary types)
+// as a list key.
+func checkForBinaryKeys(dir *ygen.ParsedDirectory) []error {
+	var errs []error
+	for _, k := range dir.ListKeys {
+		if k.LangType.NativeType == ygot.BinaryTypeName {
+			errs = append(errs, fmt.Errorf("list %s has a binary key -- this is unsupported", dir.Path))
+			continue
+		}
+		for typeName := range k.LangType.UnionTypes {
+			if typeName == ygot.BinaryTypeName {
+				errs = append(errs, fmt.Errorf("list %s has a union key containing a binary -- this is unsupported", dir.Path))
+			}
+		}
+	}
+	return errs
+}
+
+// GenerateGoCode takes a slice of strings containing the path to a set of YANG
+// files which contain YANG modules, and a second slice of strings which
+// specifies the set of paths that are to be searched for associated models (e.g.,
+// modules that are included by the specified set of modules, or submodules of those
+// modules). It extracts the set of modules that are to be generated, and returns
+// a GeneratedGoCode struct which contains:
+//	1. A struct definition for each container or list that is within the specified
+//	    set of models.
+//	2. Enumerated values which correspond to the set of enumerated entities (leaves
+//	   of type enumeration, identities, typedefs that reference an enumeration)
+//	   within the specified models.
+// If errors are encountered during code generation, an error is returned.
+func (cg *GoCodeGenerator) GenerateGoCode(yangFiles, includePaths []string) (*GeneratedGoCode, util.Errors) {
+	opts := ygen.IROptions{
+		ParseOptions:                        cg.Config.ParseOptions,
+		TransformationOptions:               cg.Config.TransformationOptions,
+		NestedDirectories:                   false,
+		AbsoluteMapPaths:                    false,
+		AppendEnumSuffixForSimpleUnionEnums: cg.GoOptions.AppendEnumSuffixForSimpleUnionEnums,
+	}
+
+	var codegenErr util.Errors
+	ir, err := ygen.GenerateIR(yangFiles, includePaths, NewGoLangMapper(cg.GoOptions.GenerateSimpleUnions), opts)
+	if err != nil {
+		return nil, util.AppendErr(codegenErr, err)
+	}
+
+	var rootName string
+	if cg.Config.TransformationOptions.GenerateFakeRoot {
+		rootName = cg.Config.TransformationOptions.FakeRootName
+		if rootName == "" {
+			rootName = igenutil.DefaultRootName
+		}
+		if r, ok := ir.Directories[fmt.Sprintf("/%s", rootName)]; ok {
+			rootName = r.Name
+		}
+	}
+	commonHeader, oneoffHeader, err := writeGoHeader(yangFiles, includePaths, cg, rootName, ir.ModelData)
```

Comment:
Should we make `writeGoHeader` just `writeHeader`?

Reply:
I didn't do this because I felt there was some clarity in answer the question "which header?"